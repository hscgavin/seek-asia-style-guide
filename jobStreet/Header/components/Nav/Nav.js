import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import styles from './nav.less';
import { Text } from 'seek-asia-style-guide/react';
import TriangleArrowDownIcon from 'seek-asia-style-guide/react/TriangleArrowDownIcon/TriangleArrowDownIcon';

const noop = i => i;

class Nav extends Component {
  static propTypes = {
    links: PropTypes.array.isRequired,
    messages: PropTypes.object.isRequired,
    activeNavLinkTextKey: PropTypes.string
  };

  constructor(props) {
    super(props);
    this.state = {
      isDropdownVisible: false
    };
    this.handleClick = this.handleClick.bind(this);
    this.handleNodeRef = this.handleNodeRef.bind(this);
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.handleClick, false);
  }

  setDropdownVisibility(isVisible) {
    const eventAction = isVisible ? 'addEventListener' : 'removeEventListener';
    document[eventAction]('click', this.handleClick, false);
    this.setState({
      isDropdownVisible: isVisible
    });
  }

  parseHrefWithHrefParams(href, hrefParams) {
    let parsedHref = href;
    if (hrefParams) {
      for (const hrefParamKey in hrefParams) {
        parsedHref = `${href}${hrefParamKey}=${hrefParams[hrefParamKey]}`;
      }
    }
    return parsedHref;
  }

  handleClick(e) {
    if (!this.dropdownNode.contains(e.target)) {
      this.setDropdownVisibility(false);
    }
  }

  handleNodeRef(node) {
    this.dropdownNode = node;
  }

  render() {
    const { links, messages, activeNavLinkTextKey } = this.props;
    const { isDropdownVisible } = this.state;
    return (
      <ul className={styles.container}>
        {links.map(link => {
          const hasChildren = Boolean(link.childLinks.length);
          const clickHandler = hasChildren ?
            e => {
              if (!isDropdownVisible) {
                this.setDropdownVisibility(true);
              }
              e.preventDefault();
            } :
            noop;

          return (
            <li
              key={link.text}
              className={classNames({
                [styles.item]: true,
                [styles.itemShowDropdown]: hasChildren && isDropdownVisible
              })}>
              <a
                className={classNames({
                  [styles.link]: true,
                  [styles.linkIsActive]: link.isActive,
                  [styles.linkHideOnMobile]: link.hideOnMobile,
                  [styles.hasRightSideDivider]: link.hasRightSideDivider
                })}
                onClick={clickHandler}
                href={this.parseHrefWithHrefParams(
                  messages[link.href] || link.href,
                  link.hrefParams
                )}
                title={messages[link.title]}>
                <Text
                  regular={link.text !== activeNavLinkTextKey}
                  whistling
                  className={styles.linkText}>
                  {link.preventTranslation ? link.text : messages[link.text]}
                  {hasChildren && (
                    <TriangleArrowDownIcon svgClassName={styles.dropdownIcon} />
                  )}
                </Text>
              </a>
              {hasChildren && (
                <ul className={styles.dropdownList} ref={this.handleNodeRef}>
                  {link.childLinks.map(childLink => (
                    <li key={childLink.text}>
                      <a
                        className={classNames({
                          [styles.dropdownLink]: true,
                          [styles.linkHideOnMobile]: childLink.hideOnMobile,
                          [styles.linkOnlyOnMobile]: childLink.onlyOnMobile,
                          [styles.hasDivider]: childLink.hasDivider
                        })}
                        href={this.parseHrefWithHrefParams(
                          messages[childLink.href],
                          childLink.hrefParams
                        )}
                        title={messages[childLink.title]}>
                        <Text regular whistling>
                          {messages[childLink.text]}
                        </Text>
                      </a>
                    </li>
                  ))}
                </ul>
              )}
            </li>
          );
        })}
      </ul>
    );
  }
}

export default Nav;
