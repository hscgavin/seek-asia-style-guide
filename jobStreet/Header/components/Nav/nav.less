@import (reference) "~seek-asia-style-guide/theme";

@linkDesktopColor: #fff;
@linkDesktopOnHoverColor: #ccc;
@secondaryLinkDesktopColor: #999;
@secondaryLinkDesktopOnHoverColor: #262626;
@linkMobileColor: #1c3f94;
@linkMobileOnHoverColor: #f5f5f5;
@itemFocusedBackgroundColor: #142d69;

.container {
  padding: @grid-base 0;
  @media @desktop {
    padding: 0;
    display: block;
  }
}

.item {
  @media @desktop {
    display: inline-block;
    color: @linkDesktopColor;
  }
}

.dropdownIcon {
  padding-top: @grid-base * 2;
  padding-left: @grid-base;
  width: @grid-base * 3;
}

.itemShowDropdown {
  & .dropdownList {
    display: block;
  }
  @media @desktop {
    background: @itemFocusedBackgroundColor;
  }
}

.link {
  display: block;
  text-decoration: none;
  color: @linkMobileColor;
  padding: @grid-base 0px;
  @media @desktop {
    padding: (@grid-base * 3) (@grid-base * 2);
    color: @linkDesktopColor;
    &:hover,
    &:focus {
      color: @linkDesktopOnHoverColor;
    }
  }
}

.dropdownList {
  display: none;
  @media @desktop {
    position: absolute;
    min-width: (@grid-base * 40);
    padding: @grid-base 0px;
    background: @saWhite;
    border: 1px solid rgba(0, 0, 0, 0.15);
    border-radius: 2px;
    box-shadow: 0 6px 12px rgba(0, 0, 0, 0.175);
    z-index: 1;
  }
}

.dropdownLink {
  display: block;
  text-decoration: none;
  color: @linkMobileColor;
  &:hover,
  &:focus {
    background: @linkMobileOnHoverColor;
    opacity: .8;
  }
  @media @desktop {
    padding: @grid-base @grid-gutter-width;
    white-space: nowrap;
    color: @secondaryLinkDesktopColor;
    &:hover,
    &:focus {
      color: @secondaryLinkDesktopOnHoverColor;
      opacity: 1;
    }
    &.hasDivider {
      border-bottom: 1px solid @linkDesktopOnHoverColor;
    }
  }
}

.linkHideOnMobile {
  display: none;
  @media @desktop {
    display: block;
  }
}

.linkOnlyOnMobile {
  display: block;
  @media @desktop {
    display: none;
  }
}
