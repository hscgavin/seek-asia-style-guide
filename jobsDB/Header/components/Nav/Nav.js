import React, {Component} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Text from 'seek-asia-style-guide/react/Text/Text';
import FbIcon from 'seek-asia-style-guide/react/FacebookIcon/FacebookIcon';
import styles from './nav.less';

const noop = i => i;

class Nav extends Component {

  static propTypes = {
    links: PropTypes.array.isRequired,
    isRightAligned: PropTypes.bool
  }
  static defaultProps = {
    isRightAligned: false,
  }

  constructor(props) {
    super(props);
    this.state = {
      isDropdownVisible: false,
    };
    this.handleClick = this.handleClick.bind(this);
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.handleClick, false);
  }

  setDropdownVisibility(isVisible) {
    const eventAction = isVisible ? 'addEventListener' : 'removeEventListener';
    document[eventAction]('click', this.handleClick, false);
    this.setState({
      isDropdownVisible: isVisible,
    });
  }

  handleClick(e) {
    if (!this.dropdownNode.contains(e.target)) {
      this.setDropdownVisibility(false);
    }
  }

  render() {
    const {links, isRightAligned} = this.props;
    const {isDropdownVisible} = this.state;
    return (
      <ul
        className={
          classNames({
            [styles.container]: true,
            [styles.isRightAligned]: isRightAligned,
          })
        }
      >
        {
          links.map((link) => {
            const hasChildren = Boolean(link.childLinks.length);
            const clickHandler = hasChildren
              ? (e) => {
                if (!isDropdownVisible) {
                  this.setDropdownVisibility(true);
                }
                e.preventDefault();
              }
              : noop;

            return (
              <li
                key={link.text}
                className={
                  classNames({
                    [styles.item]: true,
                    [styles.itemHasDropdown]: hasChildren,
                    [styles.itemShowDropdown]: hasChildren && isDropdownVisible,
                  })
                }
              >
                <a
                  className={
                    classNames({
                      [styles.link]: true,
                      [styles.linkIsActive]: link.isActive,
                      [styles.linkHideOnMobile]: link.hideOnMobile,
                    })
                  }
                  onClick={clickHandler}
                  href={'#'}
                  title={link.title}
                >
                  {
                    link.hasIcon &&
                    <FbIcon className={styles.fbIcon}/>
                  }
                  {<span>{link.text}</span>}
                </a>
                {
                  hasChildren &&
                  (
                    <ul
                      className={styles.dropdownList}
                      ref={(node) => {
                        this.dropdownNode = node;
                      }}
                    >
                      {
                        link.childLinks.map(childLink => (
                          <li
                            key={childLink.text}
                            className={styles.dropdownItem}
                          >
                            <a
                              className={styles.dropdownLink}
                              href={'#'}
                              title={'Title'}
                            >
                              <Text>{childLink.text}</Text>
                            </a>
                          </li>
                        ))
                      }
                    </ul>
                  )
                }
              </li>
            );
          })
        }
      </ul>
    );
  }
}

export default Nav;
