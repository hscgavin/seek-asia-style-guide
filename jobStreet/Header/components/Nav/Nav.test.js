import React from 'react';
import { shallow } from 'enzyme';
import Nav from './Nav';

describe('JobStreet navigation component', () => {
  it('should render supplied messages', () => {
    const links = [
      {
        href: 'unitTest.testLink',
        title: 'unitTest.testTitle',
        text: 'unitTest.testText',
        childLinks: []
      },
      {
        href: 'unitTest.testLink1',
        title: 'unitTest.testTitle1',
        text: 'unitTest.testText1',
        childLinks: []
      }
    ];
    const localization = {
      'unitTest.testLink': 'https://store.steampowered.com/?',
      'unitTest.testTitle': 'Buy SFV from Steam',
      'unitTest.testText': 'Buy now!',
      'unitTest.testLink1': 'https://store.steampowered.com/?',
      'unitTest.testTitle1': 'Buy SFV Arcade from Steam',
      'unitTest.testText1': 'Pre-order now!'
    };
    expect(
      shallow(<Nav links={links} messages={localization} />)
    ).toMatchSnapshot();
  });

  it('should render active main nav correctly', () => {
    const links = [
      {
        href: 'unitTest.testLink',
        title: 'unitTest.testTitle',
        text: 'unitTest.testText',
        childLinks: []
      },
      {
        href: 'unitTest.testLink1',
        title: 'unitTest.testTitle1',
        text: 'unitTest.testText1',
        childLinks: []
      }
    ];
    const localization = {
      'unitTest.testLink': 'https://store.steampowered.com/?',
      'unitTest.testTitle': 'Buy SFV from Steam',
      'unitTest.testText': 'Buy now!',
      'unitTest.testLink1': 'https://store.steampowered.com/?',
      'unitTest.testTitle1': 'Buy SFV Arcade from Steam',
      'unitTest.testText1': 'Pre-order now!'
    };
    expect(
      shallow(
        <Nav
          links={links}
          messages={localization}
          activeNavLinkTextKey="unitTest.testText"
        />
      )
    ).toMatchSnapshot();
  });

  it('should render message with child links', () => {
    const links = [
      {
        href: 'unitTest.testLink',
        title: 'unitTest.testTitle',
        text: 'unitTest.testText',
        childLinks: []
      }
    ];
    const localization = {
      'unitTest.testLink': 'https://store.steampowered.com/?',
      'unitTest.testTitle': 'Buy SFV from Steam',
      'unitTest.testText': 'Buy now!'
    };
    expect(
      shallow(<Nav links={links} messages={localization} isRightAligned />)
    ).toMatchSnapshot();
  });

  it('should render message with URL parameters', () => {
    const links = [
      {
        href: 'unitTest.testLink',
        hrefParams: {
          sfvFavorite: 'karin'
        },
        title: 'unitTest.testTitle',
        text: 'unitTest.testText',
        childLinks: []
      }
    ];
    const localization = {
      'unitTest.testLink': 'https://store.steampowered.com/?',
      'unitTest.testTitle': 'Buy SFV from Steam',
      'unitTest.testText': 'Buy now!'
    };
    expect(
      shallow(<Nav links={links} messages={localization} />)
    ).toMatchSnapshot();
  });
});
