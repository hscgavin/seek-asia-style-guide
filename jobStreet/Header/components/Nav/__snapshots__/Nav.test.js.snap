// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobStreet navigation component should render active main nav correctly 1`] = `
<ul
  className="container"
>
  <li
    className="item"
    key="unitTest.testText"
  >
    <a
      className="link"
      href="https://store.steampowered.com/?"
      onClick={[Function]}
      title="Buy SFV from Steam"
    >
      <Text
        baseline={true}
        className="linkText"
        regular={false}
        whistling={true}
      >
        Buy now!
      </Text>
    </a>
  </li>
  <li
    className="item"
    key="unitTest.testText1"
  >
    <a
      className="link"
      href="https://store.steampowered.com/?"
      onClick={[Function]}
      title="Buy SFV Arcade from Steam"
    >
      <Text
        baseline={true}
        className="linkText"
        regular={true}
        whistling={true}
      >
        Pre-order now!
      </Text>
    </a>
  </li>
</ul>
`;

exports[`JobStreet navigation component should render message with URL parameters 1`] = `
<ul
  className="container"
>
  <li
    className="item"
    key="unitTest.testText"
  >
    <a
      className="link"
      href="https://store.steampowered.com/?sfvFavorite=karin"
      onClick={[Function]}
      title="Buy SFV from Steam"
    >
      <Text
        baseline={true}
        className="linkText"
        regular={true}
        whistling={true}
      >
        Buy now!
      </Text>
    </a>
  </li>
</ul>
`;

exports[`JobStreet navigation component should render message with child links 1`] = `
<ul
  className="container"
>
  <li
    className="item"
    key="unitTest.testText"
  >
    <a
      className="link"
      href="https://store.steampowered.com/?"
      onClick={[Function]}
      title="Buy SFV from Steam"
    >
      <Text
        baseline={true}
        className="linkText"
        regular={true}
        whistling={true}
      >
        Buy now!
      </Text>
    </a>
  </li>
</ul>
`;

exports[`JobStreet navigation component should render supplied messages 1`] = `
<ul
  className="container"
>
  <li
    className="item"
    key="unitTest.testText"
  >
    <a
      className="link"
      href="https://store.steampowered.com/?"
      onClick={[Function]}
      title="Buy SFV from Steam"
    >
      <Text
        baseline={true}
        className="linkText"
        regular={true}
        whistling={true}
      >
        Buy now!
      </Text>
    </a>
  </li>
  <li
    className="item"
    key="unitTest.testText1"
  >
    <a
      className="link"
      href="https://store.steampowered.com/?"
      onClick={[Function]}
      title="Buy SFV Arcade from Steam"
    >
      <Text
        baseline={true}
        className="linkText"
        regular={true}
        whistling={true}
      >
        Pre-order now!
      </Text>
    </a>
  </li>
</ul>
`;
